### PR TITLE
Avoid generic interface GetEnumerator empty allocations in immutable collections

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/corefx/issues/24563
cc: @davkean, @AArnott

### CODE REVIEW

**Comment by stephentoub** [comment]
@dotnet/dnceng, this `error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun` error is happening again on OSX x64 Debug Build (https://ci3.dot.net/job/dotnet_corefx/job/master/job/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/4187/).  It looks like it's the same machine that was failing before.

@dotnet-bot test OSX x64 Debug Build please

**Comment by AArnott** [comment]
Why make the array enumerator `internal` for the sake of an empty enumerator when you can just use `Enumerable.Empty<T>()`? IIRC that is a zero-allocation version as well.

For the Builders, have you confirmed that this isn't a publicly observable breaking change? For instance, if I call GetEnumerator on an empty Builder today (before your change), and then I populate that builder, do all the Builders continue to produce zero elements from that enumerator or might the enumerator show the _current_ contents of the Builder? I'd be concerned if your change causes some special behavior where it "locks in" an empty enumerator when it is obtained from an empty Builder but in any other state, the Builder produces a "live" Enumerator.

**Comment by stephentoub** [comment]
> Why make the array enumerator internal for the sake of an empty enumerator when you can just use Enumerable.Empty<T>()?

Because I saw that ImmutableArray was using its own custom one instead of Enumerable.Empty, figured there was a reason (maybe it's trying to avoid the allocation on versions of .NET before array implemented the singleton enumerator for 0-length arrays?), and just used it.  If there's no such reason, I'm happy to switch to Enumerable.Empty.

> For the Builders, have you confirmed that this isn't a publicly observable breaking change?

It is observable, in that if you get the enumerator for the empty builder, then add to the builder, and then continue enumerating, it'll throw a collection modified exception, and with this change it won't.  If that's concerning, I can roll back the change to the builders.

**Comment by AArnott** [comment]
> If there's no such reason, I'm happy to switch to Enumerable.Empty.

If that reason existed, I think it's moot now. This library used to target .NET 3.5 and 4.0, but no longer. So I think all supported target frameworks have that optimization at this point.

> If that's concerning, I can roll back the change to the builders.

Interesting. I'm glad we throw like we should there. :) As for this change to the Builders, I'm leaning toward not having the special case for more predictable behavior. Also once we're enumerating Builders I think the memory pressure is perhaps less important than the immutable ones. But I don't know what drove this PR in the first place. So if you can achieve your goal without changing the Builders, I think I'm more comfortable with that, thanks.

**Comment by stephentoub** [comment]
> If that reason existed, I think it's moot now

Ok, switched to Enumerable.Empty.

> I'm leaning toward not having the special case for more predictable behavior

Removed the builder changes.

> I don't know what drove this PR in the first place

The linked issue David filed.  As long as I was making the change for dictionary, I applied it elsewhere, too.

**Comment by stephentoub** [comment]
@dotnet-bot test Windows x64 Debug Build please (https://github.com/dotnet/corefx/issues/23341)

**Comment by stephentoub** [comment]
> May as well fix the IEnumerable.GetEnumerator() case at the same time.

Is one extra potential allocation really a concern if you're using the non-generic interface?

**Comment by sharwell** [comment]
> Extra code has cost; even if small, is it really worth it?

Just make `IEnumerable.GetEnumerator()` call `((IEnumerable<T>)this).GetEnumerator()` instead of `this.GetEnumerator()`.

**Comment by stephentoub** [comment]
> Just make IEnumerable.GetEnumerator() call ((IEnumerable<T>)this).GetEnumerator() instead of this.GetEnumerator().

I'm not saying I don't know how to make the change.  I'm saying, why bother?  Any churn has risk, so there needs to be a reason.  What's the value of doing so?  Can you point me to a single case in Roslyn, in Visual Studio, in any code that cares about performance where it's iterating an immutable collection using the non-generic interface?  And if so, how much could it really care about performance?

**Comment by sharwell** [comment]
I cannot, so I'll change my ❕ to :bulb:

**Comment by mmitche** [comment]
@stephentoub I literally turned that machine off yesterday....how did it get back on...

**Comment by stephentoub** [comment]
>  I literally turned that machine off yesterday....how did it get back on...

Seems like the beginning of a bad B horror movie. Revenge of the living dead macs.

**Comment by mmitche** [comment]
> Seems like the beginning of a bad B horror movie. Revenge of the living dead macs.

IT CAME BAAAACKKKKK! IT CAME BACKKK!

**Comment by stephentoub** [comment]
"Just when you thought it was safe to go back in the cloud.... reee, reee, reee..."

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs [hunk 1]
```diff
@@ -920,7 +921,10 @@ IImmutableList<T> IImmutableList<T>.InsertRange(int index, IEnumerable<T> items)
         /// <returns>
         /// A <see cref="IEnumerator{T}"/> that can be used to iterate through the collection.
         /// </returns>
-        IEnumerator<T> IEnumerable<T>.GetEnumerator() => this.GetEnumerator();
+        IEnumerator<T> IEnumerable<T>.GetEnumerator() =>
```
Comment (AArnott) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs]: nit: I prefer to use the expression-bodied methods for one-liners, and this isn't one-line any more.

Comment (stephentoub) [reply src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs]: Changed

**Pull request merged**